src/tableofcontentbar.cpp

```diff
@@ -0,0 +1,85 @@
+#include "tableofcontentbar.h"
+#include "ui_tableofcontentbar.h"
+#include "kiwixapp.h"
+#include <QJsonObject>
+#include <QTreeWidgetItem>
+
+TableOfContentBar::TableOfContentBar(QWidget *parent) :
+    QFrame(parent),
+    ui(new Ui::tableofcontentbar)
+{
+    ui->setupUi(this);
+    ui->titleLabel->setFont(QFont("Selawik", 18, QFont::Weight::Medium));
+    ui->titleLabel->setText(gt("table-of-content"));
+    ui->hideLabel->setFont(QFont("Selawik", 12));
+    ui->hideLabel->setTextFormat(Qt::RichText);
+
+    /* href is needed to make hide clickable, but not used. So Kiwix it is :) */
+    ui->hideLabel->setText("<a href=\"https://kiwix.org/en/\">" + gt("hide") + "</a>");
+    connect(ui->hideLabel, &QLabel::linkActivated, this, [=](){
+        KiwixApp::instance()->getAction(KiwixApp::ToggleTOCAction)->setChecked(false);
+    });
+
+    ui->tree->setRootIsDecorated(false);
+    ui->tree->setItemsExpandable(false);
+    connect(ui->tree, &QTreeWidget::itemClicked, this, &TableOfContentBar::onTreeItemActivated);
+    connect(ui->tree, &QTreeWidget::itemActivated, this, &TableOfContentBar::onTreeItemActivated);
+}
+
+TableOfContentBar::~TableOfContentBar()
+{
+    delete ui;
+}
+
+void TableOfContentBar::onTreeItemActivated(QTreeWidgetItem *item)
+{
+    emit navigationRequested(m_url, item->data(0, Qt::UserRole).toString());
+}
+
+namespace
+{
+
+void populateItem(QJsonArray& headerArr, QTreeWidgetItem* parent, int level)
+{
+    while (!headerArr.isEmpty())
+    {
+        const auto header = headerArr.first().toObject();
+        const int headerLevel = header["level"].toInt();
+        if (headerLevel <= level)
+            return;
+        else
+            headerArr.pop_front();
+
+        const auto item = new QTreeWidgetItem(parent);
+        item->setExpanded(true);
+
+        auto numberList = parent->data(0, Qt::UserRole + 1).toStringList();
+        numberList.append(QString::number(parent->childCount()));
+        item->setData(0, Qt::UserRole + 1, numberList);
+
+        const auto itemNum = numberList.join(".");
+        const auto display = itemNum + "  " + header["text"].toString();
+        item->setData(0, Qt::DisplayRole, display);
+        item->setData(0, Qt::FontRole, QFont("Selawik", 12));
+        item->setData(0, Qt::UserRole, header["anchor"].toString());
+        item->setToolTip(0, display);
+        populateItem(headerArr, item, headerLevel);
+    }
+}
```

Comment:
My intuition tells me that this combination of iteration and recursion is redundant (and to my taste a little confusing). I feel that this algorithm can be rewritten in a clearer way with iteration alone (you will need an additional nested loop for ascending up the hierarchy to the `QTreeWidgetItem` corresponding to the new level). To add more clarity please extract the code responsible for the creation and initialization of a new `QTreeWidgetItem` into a helper function of its own.

Reply:
Done